media-proxy/src/rdma_session.c

```diff
@@ -0,0 +1,460 @@
+    }
+
+    // TODO: use memif buffer with correct size
+    ep_cfg_t ep_cfg = {
+        .rdma_ctx = tx_ctx->rdma_ctx,
+        .data_buf_size = tx_ctx->transfer_size,
+        .local_addr = opts->local_addr,
+        .remote_addr = opts->remote_addr,
+        .data_buf = malloc(tx_ctx->transfer_size),
+        .dir = opts->dir,
+    };
+    if (!ep_cfg.data_buf) {
+        printf("%s, session data buffer malloc fail\n", __func__);
+        return NULL;
+    }
+
+    ret = ep_init(&tx_ctx->ep_ctx, &ep_cfg);
+    if (ret) {
+        printf("%s, fail to initialize libfabric's end point.\n", __func__);
+        free(tx_ctx);
+        return NULL;
+    }
+
+    return tx_ctx;
+}
+
+static void rx_rdma_consume_frame(rx_rdma_session_context_t* s, char* frame)
+{
+    int err = 0;
+    uint16_t qid = 0;
+    mcm_buffer* rx_mcm_buff = NULL;
+    memif_buffer_t* rx_bufs = NULL;
+    uint16_t buf_num = 1;
+    memif_conn_handle_t conn;
+    uint32_t buf_size = s->transfer_size;
+    uint16_t rx_buf_num = 0, rx = 0;
+
+    if (s->shm_ready == 0) {
+        INFO("%s memif not ready\n", __func__);
+        return;
+    }
+
+    rx_bufs = s->shm_bufs;
+
+    /* allocate memory */
+    err = memif_buffer_alloc(s->memif_conn, qid, rx_bufs, buf_num, &rx_buf_num, buf_size);
+    if (err != MEMIF_ERR_SUCCESS) {
+        INFO("rx_rdma_consume_frame: Failed to alloc memif buffer: %s", memif_strerror(err));
+        return;
+    }
+
+    memcpy(rx_bufs->data, frame, s->transfer_size);
+
+    /* Send to microservice application. */
+    err = memif_tx_burst(s->memif_conn, qid, rx_bufs, rx_buf_num, &rx);
+    if (err != MEMIF_ERR_SUCCESS) {
+        INFO("rx_rdma_consume_frame memif_tx_burst: %s", memif_strerror(err));
+    }
+
+    s->fb_recv++;
+}
+
+static void* rx_rdma_frame_thread(void* arg)
+{
+    rx_rdma_session_context_t* s_ctx = (rx_rdma_session_context_t*)arg;
+    libfabric_ctx* rdma_ctx = s_ctx->rdma_ctx;
+    ep_ctx_t* cp_ctx = s_ctx->ep_ctx;
+
+    printf("%s(%d), start\n", __func__, s_ctx->idx);
+    while (!s_ctx->stop) {
+        ep_recv_buf(cp_ctx, cp_ctx->data_buf ,s_ctx->transfer_size);
+        rx_rdma_consume_frame(s_ctx, cp_ctx->data_buf);
+    }
+
+    return NULL;
+}
+
+/* RX: Create RDMA session */
+rx_rdma_session_context_t* rdma_rx_session_create(libfabric_ctx* dev_handle, rdma_s_ops_t* opts, memif_ops_t* memif_ops){
+    int ret = 0;
+	rx_rdma_session_context_t* rx_ctx;
+    rx_ctx = calloc(1, sizeof(rx_rdma_session_context_t));
+    if (rx_ctx == NULL) {
+        printf("%s, TX session contex malloc fail\n", __func__);
+        return rx_ctx;
+    }
+    rx_ctx->rdma_ctx = dev_handle;
+    rx_ctx->stop = false;
+
+    rx_ctx->transfer_size = opts->transfer_size;
+
+    ret = rx_rdma_shm_init(rx_ctx, memif_ops);
+
+    if (ret < 0) {
+        printf("%s, fail to initialize share memory.\n", __func__);
+        free(rx_ctx);
+        return NULL;
+    }
+
+    // TODO: use memif buffer with correct size
+    ep_cfg_t ep_cfg = {
```

Comment:
Variable declaration shouldn't be done in the middle of the code in C. Move it to the beginning of the block.

Reply:
fixed